In [1]:
!pip install spacy


   ---------------------------------------- 0.0/11.8 MB ? eta -:--:--
   - -------------------------------------- 0.5/11.8 MB 5.7 MB/s eta 0:00:02
   ------------------- -------------------- 5.8/11.8 MB 20.8 MB/s eta 0:00:01
   ---------------------------------------- 11.8/11.8 MB 25.4 MB/s eta 0:00:00
   ---------------------------------------- 0.0/632.6 kB ? eta -:--:--
   --------------------------------------- 632.6/632.6 kB 24.7 MB/s eta 0:00:00
   ---------------------------------------- 0.0/1.5 MB ? eta -:--:--
   ---------------------------------------- 1.5/1.5 MB 37.3 MB/s eta 0:00:00
   ---------------------------------------- 0.0/6.3 MB ? eta -:--:--
   ------------------------------- -------- 5.0/6.3 MB 27.4 MB/s eta 0:00:01
   ---------------------------------------- 6.3/6.3 MB 20.2 MB/s eta 0:00:00
   ---------------------------------------- 0.0/5.4 MB ? eta -:--:--
   ---------------------------------------- 5.4/5.4 MB 36.3 MB/s eta 0:00:00


In [3]:
!python -m spacy download en_core_web_sm


     ---------------------------------------- 0.0/12.8 MB ? eta -:--:--
      --------------------------------------- 0.3/12.8 MB ? eta -:--:--
     ------------- -------------------------- 4.2/12.8 MB 16.8 MB/s eta 0:00:01
     --------------------------------------- 12.8/12.8 MB 27.7 MB/s eta 0:00:00
[+] Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


In [18]:
import spacy
nlp = spacy.load("en_core_web_sm")
print("SpaCy and the language model are working")
# Just to cheke it is workning


SpaCy and the language model are working


# AS1

# lemmatization

In [3]:
import spacy


def read_file(file_path):
    """
    Read the contents of a text file.

    Args:
        file_path (str): The path to the file to be read.

    Returns:
        str: The contents of the file as a single string.
    """
    with open(file_path, "r", encoding="utf-8") as file:
        return file.read()


def lemmatize_text(text):
    """
    Lemmatize the input text using spaCy's language model.

    Args:
        text (str): The original text to lemmatize.

    Returns:
        list: A list of lemmatized words from the text.
    """
    nlp = spacy.load("en_core_web_sm")
    doc = nlp(text)
    return [token.lemma_ for token in doc]


def save_to_file(words, file_path):
    """
    Save a list of words to a file.

    Args:
        words (list): The list of words to save.
        file_path (str): The path to the output file.
    """
    with open(file_path, "w", encoding="utf-8") as file:
        file.write(" ".join(words))


def count_changed_words(original, lemmatized):
    """
    Count how many words changed after lemmatization.

    Args:
        original (list): The original list of words.
        lemmatized (list): The lemmatized list of words.

    Returns:
        int: The number of words that changed after lemmatization.
    """
    return sum(1 for orig, lemma in zip(original, lemmatized) if orig != lemma)


def main():
    """
    Main function to read, lemmatize, and save text,
    and display the total number of changed words.
    """
    # File paths
    input_file = "wiki_mountain_def.txt"
    output_file = "lemmatized_output.txt"

    # Process
    text = read_file(input_file)
    original_words = text.split()
    lemmatized_words = lemmatize_text(text)
    save_to_file(lemmatized_words, output_file)

    # Count and report changes
    changed_words = count_changed_words(original_words, lemmatized_words)
    print(f"Lemmatization complete. Total words changed to : {changed_words}")

In [5]:
if __name__ == "__main__":
    main()

Lemmatization complete. Total words changed to : 225


# Stemming Code

In [7]:
def porter_stem(word):
    """
    Apply a subset of the Porter stemming rules.

    Rules applied:
    - SSES -> SS  (e.g., caresses → caress)
    - IES -> I    (e.g., ponies → poni)
    - SS -> SS    (e.g., caress → caress)
    - S -> (remove) (e.g., cats → cat)

    Args:
        word (str): The word to be stemmed.

    Returns:
        str: The stemmed word according to the specified rules.
    """
    if word.endswith("sses"):
        return word[:-2]  # Remove 'es' from 'sses'
    elif word.endswith("ies"):
        return word[:-3] + "i"  # Replace 'ies' with 'i'
    elif word.endswith("ss"):
        return word  # Leave 'ss' unchanged
    elif word.endswith("s"):
        return word[:-1]  # Remove the trailing 's'
    else:
        return word


def perform_stemming(input_file, output_file):
    """
    Read text from a file, apply stemming rules, save the output, and
    count how many words were changed.

    Args:
        input_file (str): Path to the original text file.
        output_file (str): Path to save the stemmed output.

    Returns:
        int: Total number of words changed during stemming.
    """
    # Read the original text
    with open(input_file, "r", encoding="utf-8") as file:
        text = file.read()

    original_words = text.split()
    stemmed_words = [porter_stem(word) for word in original_words]

    # Save the stemmed words to a file
    with open(output_file, "w", encoding="utf-8") as f:
        f.write(" ".join(stemmed_words))

    # Count the number of words that changed after stemming
    changed_count = sum(1 for orig, stem in zip(original_words, stemmed_words) if orig != stem)
    return changed_count


def main():
    """
    Main function to perform stemming and display the number of changed words.
    """
    input_file = "wiki_mountain_def.txt"
    output_file = "stemmed_output.txt"

    changed_words = perform_stemming(input_file, output_file)
    print(f"Stemming complete. Total words changed to: {changed_words}")

In [9]:
if __name__ == "__main__":
    main()


Stemming complete. Total words changed to: 31


# differences between lemmatization and stemming results

# Two important methods in natural language processing (NLP) for breaking down words into their basic forms are lemmatization and stemming.  Despite having a similar objective, their approaches to achieving it are very different.  By considering the context and part of speech, lemmatization converts words into their fundamental dictionary form, or lemma.  For instance, "better" would become "good," while "running" would lemmatize to "run."  By doing this, the words' actual meanings are preserved.  Lemmatization is more accurate but may take longer to process because it uses complex language models like SpaCy.

# In contrast, stemming removes suffixes from words using straightforward rule-based algorithms such as the Porter Stemmer.  This approach may result in errors since it ignores the context of the word.  For example, "ponies" may be mispronounced as "poni," even while "running" is appropriately reduced to "run."  Although stemming is quicker and more effective, accuracy is frequently sacrificed, and it occasionally creates word forms that don't exist.

# Lemmatization is the superior option in real-world applications where meaning and accuracy are crucial, like in text analysis or machine learning models.  Stemming, however, works better in situations where speed is essential, such as search engines, when a small amount of error is acceptable.  Lemmatization altered 225 words in the assignment results, demonstrating its more thorough and accurate processing.  Stemming, on the other hand, only altered 31 words, demonstrating its more straightforward, rule-based methodology.  The decision between lemmatization and stemming ultimately comes down to whether the task requires quick processing or great accuracy.